File: fsharp/tournament/Tournament.fs
module Tournament

let tally input = 
    let update team res teams =
        let (mp,w,d,l,p) = match Map.tryFind team teams with | Some n -> n | None -> (0,0,0,0,0)
        let newpoints = 
            match res with
            | "win" -> (mp + 1,w + 1,d,l,p + 3)
            | "draw" -> (mp + 1,w,d + 1,l,p + 1)
            | _ -> (mp + 1,w,d,l + 1,p)
        teams.Add (team, newpoints)
    let rev res = match res with | "win" -> "loss" | "loss" -> "win" | _ -> res

    let result = 
        input 
        |> List.map (fun (l:string) -> l.Split(';') |> fun l -> (l.[0],l.[1],l.[2]))
        |> List.fold 
            (fun teams (t1, t2, res) -> 
                let teams2 = update t1 res teams
                update t2 (rev res) teams2) Map.empty<string, (int * int * int * int * int)>
        |> Map.toList 
        |> List.sortBy (fun (team,(_,_,_,_,_)) -> [team]) |> List.sortByDescending (fun (_,(_,_,_,_,p)) -> [p])

    [sprintf "%-31s| MP |  W |  D |  L |  P" "Team"] 
        @ (result |> List.map (fun (team,(mp,w,l,d,p)) -> sprintf "%-31s| %2i | %2i | %2i | %2i | %2i" team mp w l d p))

File: go/tournament/tournament.go
package tournament

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"sort"
	"strings"
)

type team struct {
	wins, losses, draws int
}

func (team team) Score() int {
	return team.wins*3 + team.draws
}

// Tally calculates the scores for all teams from a match list input
func Tally(allMatches io.Reader, allResults io.Writer) error {
	matchReader := bufio.NewScanner(allMatches)
	teams := map[string]*team{}

	for matchReader.Scan() {
		line := matchReader.Text()
		if line == "" || line[0] == '#' {
			continue
		}

		aspects := strings.Split(line, ";")
		if len(aspects) != 3 {
			return errors.New("invalid line")
		}

		teamA, teamB, result := aspects[0], aspects[1], aspects[2]
		if _, ok := teams[teamA]; !ok {
			teams[teamA] = &team{}
		}
		if _, ok := teams[teamB]; !ok {
			teams[teamB] = &team{}
		}

		if result == "win" {
			teams[teamA].wins++
			teams[teamB].losses++
		} else if result == "draw" {
			teams[teamA].draws++
			teams[teamB].draws++
		} else if result == "loss" {
			teams[teamA].losses++
			teams[teamB].wins++
		} else {
			return errors.New("invalid line")
		}
	}

	_, err := fmt.Fprintf(allResults, "%-31s| MP |  W |  D |  L |  P\n", "Team")
	if err != nil {
		return err
	}

	for _, key := range sortedKeys(teams) {
		team := teams[key]
		mp := team.wins + team.losses + team.draws
		_, err := fmt.Fprintf(allResults, "%-31s|  %d |  %d |  %d |  %d |  %d\n",
			key, mp, team.wins, team.draws, team.losses, team.Score())
		if err != nil {
			return err
		}
	}

	return nil
}

func sortedKeys(teams map[string]*team) []string {
	var keys []string
	for key := range teams {
		keys = append(keys, key)
	}

	sort.Slice(keys, func(a, b int) bool {
		scoreA := teams[keys[a]].Score()
		scoreB := teams[keys[b]].Score()
		if scoreA == scoreB {
			return keys[a] < keys[b]
		}
		return scoreA > scoreB
	})

	return keys
}
